src/sys/socket/sockopt.rs

```diff
@@ -1001,6 +1001,133 @@ sockopt_impl!(
+    libc::c_uint
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Drop socket from group
+    NetlinkDropMembership,
+    SetOnly,
+    libc::SOL_NETLINK,
+    libc::NETLINK_DROP_MEMBERSHIP,
+    libc::c_uint
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Enable packet info control messages
+    NetlinkPacketInfo,
+    Both,
+    libc::SOL_NETLINK,
+    libc::NETLINK_PKTINFO,
+    bool
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Send broadcast errors
+    NetlinkBroadcastError,
+    Both,
+    libc::SOL_NETLINK,
+    libc::NETLINK_BROADCAST_ERROR,
+    bool
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Avoid receiving 'ENOBUFS' errors
+    NetlinkNoEnobufs,
+    Both,
+    libc::SOL_NETLINK,
+    libc::NETLINK_NO_ENOBUFS,
+    bool
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Listen to notifications from all network namespaces
+    NetlinkListenAllNsid,
+    SetOnly,
+    libc::SOL_NETLINK,
+    libc::NETLINK_LISTEN_ALL_NSID,
+    bool
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Trim the extack if needed
+    NetlinkCapAck,
+    Both,
+    libc::SOL_NETLINK,
+    libc::NETLINK_CAP_ACK,
+    bool
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Always send extack
+    NetlinkExtAck,
+    Both,
+    libc::SOL_NETLINK,
+    libc::NETLINK_EXT_ACK,
+    bool
+);
+#[cfg(any(target_os = "android", target_os = "linux"))]
+sockopt_impl!(
+    /// Get strict check
+    NetlinkGetStrictCheck,
+    Both,
+    libc::SOL_NETLINK,
+    libc::NETLINK_GET_STRICT_CHK,
+    bool
+);
+
+/// List netlink groups
+/// The returned Vec contains a bit set in chunks of u32.
+#[cfg(any(target_os = "android", target_os = "linux"))]
+#[derive(Copy, Clone, Debug)]
+pub struct NetlinkListMemberships;
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
+impl GetSockOpt for NetlinkListMemberships {
+    type Val = Vec<u32>;
+
+    fn get(&self, fd: RawFd) -> Result<Self::Val> {
+        let mut len = 0;
+        unsafe {
+            // Get membership length
+            let res = libc::getsockopt(
+                fd,
+                libc::SOL_NETLINK,
+                libc::NETLINK_LIST_MEMBERSHIPS,
+                ::std::ptr::null_mut(),
+                &mut len,
+            );
+            Errno::result(res)?;
+            let mut val = vec![0u32; (len as usize) / mem::size_of::<u32>()];
+            if val.capacity() > 0 {
+                let old_len = len;
+                let res = libc::getsockopt(
+                    fd,
+                    libc::SOL_NETLINK,
+                    libc::NETLINK_LIST_MEMBERSHIPS,
+                    val.as_mut_ptr() as *mut libc::c_void,
+                    &mut len,
+                );
+                if old_len != len {
```

Comment:
This behavior isn't correct.  If `old_len != len`, that probably indicates that the group membership changed after the first call to `getsockopt`.  You shouldn't return an error in that case.  If the group membership has shrunk, then you should adjust the size of the vec accordingly.  If the list has grown, then you should extend the size of the Vec and try again.
Also, you don't need to zero-initialize the vec.  You can do `Vec::with_capacity` and allow `getsockopt` to initialize it.  See `unistd::getgrouplist` for an example of that pattern.

Reply:
> This behavior isn't correct. If `old_len != len`, that probably indicates that the group membership changed after the first call to `getsockopt`. 
Correct.
> You shouldn't return an error in that case. If the group membership has shrunk, then you should adjust the size of the vec accordingly. If the list has grown, then you should extend the size of the Vec and try again.
ACK

Reply:
Am looping back when `old_len != len`, so there's extremely low chance for an endless loop if the groups are changed by some buggy code. Should we add a try counter?